services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/CompletedDownload.java

```diff
@@ -15,33 +15,52 @@
 
 package software.amazon.awssdk.transfer.s3;
 
+import java.util.Objects;
 import software.amazon.awssdk.annotations.SdkPreviewApi;
 import software.amazon.awssdk.annotations.SdkPublicApi;
-import software.amazon.awssdk.services.s3.S3AsyncClient;
-import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.transfer.s3.CompletedDownload.TypedBuilder;
+import software.amazon.awssdk.utils.ToString;
 import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
 
 /**
- * Represents a completed download transfer from Amazon S3. It can be used to track
- * the underlying {@link GetObjectResponse}
+ * Represents a completed download transfer from Amazon S3. It can be used to track the underlying result
+ * that was transformed via an {@link AsyncResponseTransformer}.
  *
- * @see S3TransferManager#download(DownloadRequest)
+ * @see S3TransferManager#download(DownloadRequest) 
  */
 @SdkPublicApi
 @SdkPreviewApi
-public final class CompletedDownload implements CompletedTransfer {
-    private final GetObjectResponse response;
+public final class CompletedDownload<ResultT>
+    implements CompletedObjectTransfer,
+               ToCopyableBuilder<TypedBuilder<ResultT>, CompletedDownload<ResultT>> {
 
-    private CompletedDownload(DefaultBuilder builder) {
-        this.response = Validate.paramNotNull(builder.response, "response");
+    private final ResultT result;
+
+    private CompletedDownload(DefaultTypedBuilder<ResultT> builder) {
+        this.result = Validate.paramNotNull(builder.result, "result");
     }
 
+
     /**
-     * Returns the API response from the {@link S3TransferManager#download(DownloadRequest)}
-     * @return the response
+     * Create a builder that can be used to create a {@link CompletedDownload}.
+     *
+     * @see UntypedBuilder
      */
-    public GetObjectResponse response() {
-        return response;
+    public static UntypedBuilder builder() {
+        return new DefaultUntypedBuilder();
+    }
```

Comment:
Our standard for typed builders is to accept the value class or value in the builder() method call: https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/model/PutItemEnhancedResponse.java#L69-L71

Reply:
Do we need to create the `CompletedDownload` before we have the `ResponseBytes<GetObjectResponse>`? If not, we can just pass that into the `builder` method.

I guess we could also do `CompletedDownload.<ResponseBytes<GetObjectResponse>>builder()` to initialize the type parameters at creation time, if we absolutely have to. It's not particularly intuitive for customers to do, but if they don't need to do it much (since it's primarily a response type) it's simpler than the mixed builder approach (imo).

Reply:
> Do we need to create the CompletedDownload before we have the ResponseBytes<GetObjectResponse>? If not, we can just pass that into the builder method.

We do need to specify the type. The current flow looks like:
1. `DownloadRequest<T>` contains an `AsyncResponseTransformer<GetObjectResponse,T>`
1. `Download<T>` contains a `CompleteableFuture<CompletedDownload<T>>`
1. `CompletedDownload<T>` contains a `T`

> I guess we could also do CompletedDownload.<ResponseBytes<GetObjectResponse>>builder() to initialize the type parameters at creation time, if we absolutely have to. It's not particularly intuitive for customers to do, but if they don't need to do it much (since it's primarily a response type) it's simpler than the mixed builder approach (imo).

We can definitely use that method, but that type of syntax is what I was explicitly trying to avoid, since I think most users will find it very cumbersome and unfamiliar. While the mixed builder approach is more complexity on our end, most users don't need to be concerned with it if they're declaring their builders in one sweep.

Typical user code with a single, generic builder:

```
Download<ResponseBytes<GetObjectResponse>> download =
    tm.download(DownloadRequest.<ResponseBytes<GetObjectResponse>>builder()
                               .getObjectRequest(b -> b.bucket(BUCKET).key(KEY))
                               .responseTransformer(AsyncResponseTransformer.toBytes())
                               .build());
```

Typical user code with the mixed builder:

```
Download<ResponseBytes<GetObjectResponse>> download =
    tm.download(DownloadRequest.builder()
                               .getObjectRequest(b -> b.bucket(BUCKET).key(KEY))
                               .responseTransformer(AsyncResponseTransformer.toBytes())
                               .build());
```

In this case, it only simplifies things from a user perspective.

Reply:
I also tried to document this behavior here:

https://github.com/aws/aws-sdk-java-v2/blob/9f8e42b08a34cc32d7ba7ac92b340be8153de50c/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/DownloadRequest.java#L132-L142